webui/react/src/pages/TrialDetails/LogViewer.tsx

```diff
@@ -0,0 +1,527 @@
+  const formattedTime = log.time ? formatDatetime(log.time, { format: DATETIME_FORMAT }) : '';
+  return { ...log, formattedTime };
+};
+
+const formatClipboardHeader = (log: Log): string => {
+  const logEntry = formatLogEntry(log);
+  const format = `%${MAX_DATETIME_LENGTH - 1}s `;
+  const level = `<${logEntry.level || ''}>`;
+  return sprintf(`%-9s ${format}`, level, logEntry.formattedTime);
+};
+
+const logSorter =
+  (key: keyof Log) =>
+  (a: Log, b: Log): number => {
+    const aValue = a[key];
+    const bValue = b[key];
+    if (key === 'id') return numericSorter(aValue as number, bValue as number);
+    if (key === 'time') return dateTimeStringSorter(aValue as string, bValue as string);
+    return 0;
+  };
+
+// This is an arbitrarily large number used as an index. See https://virtuoso.dev/prepend-items/
+const START_INDEX = 2_000_000_000;
+
+function LogViewer<T>({
+  decoder,
+  height = 'stretch',
+  initialLogs,
+  onDownload,
+  onFetch,
+  onError,
+  serverAddress,
+  sortKey = 'time',
+  handleCloseLogs,
+  title,
+  selectedLog,
+}: Props<T>): JSX.Element {
+  const componentId = useId();
+  const logsRef = useRef<HTMLDivElement>(null);
+  const virtuosoRef = useRef<VirtuosoHandle>(null);
+  const [isFetching, setIsFetching] = useState(false);
+  const local = useRef({
+    idSet: new Set<RecordKey>([]),
+    isScrollReady: false as boolean,
+  });
+  const [canceler] = useState(new AbortController());
+  const [fetchDirection, setFetchDirection] = useState<FetchDirection>(FetchDirection.Older);
+  const [isTailing, setIsTailing] = useState<boolean>(true);
+  const [showButtons, setShowButtons] = useState(false);
+  const [logs, setLogs] = useState<ViewerLog[]>([]);
+  const [firstItemIndex, setFirstItemIndex] = useState(START_INDEX);
+  const [scrolledForSearch, setScrolledForSearch] = useState(true);
+
+  const processLogs = useCallback(
+    (newLogs: Log[]) => {
+      return newLogs
+        .filter((log) => {
+          const isDuplicate = local.current.idSet.has(log.id);
+          const isTqdm = log.message.includes('\r');
+          local.current.idSet.add(log.id);
+          return !isDuplicate && !isTqdm;
+        })
+        .map((log) => formatLogEntry(log))
+        .sort(logSorter(sortKey));
+    },
+    [sortKey],
+  );
+
+  const addLogs = useCallback((newLogs: ViewerLog[] = [], prepend = false): void => {
+    if (newLogs.length === 0) return;
+    setLogs((prevLogs) => (prepend ? newLogs.concat(prevLogs) : prevLogs.concat(newLogs)));
+    if (prepend) setFirstItemIndex((prev) => prev - newLogs.length);
+  }, []);
+
+  useEffect(() => {
+    if (scrolledForSearch || !selectedLog || !local.current.isScrollReady) return;
+    setTimeout(() => {
+      virtuosoRef.current?.scrollToIndex({
+        behavior: 'smooth',
+        index: logs.findIndex((l) => l.id === selectedLog.id),
+      });
+      if (logs.findIndex((l) => l.id === selectedLog.id) >= 0) setScrolledForSearch(true);
+    }, 250);
```

Comment:
it might take more lines, but saving the index allows us to guarantee that we don't scroll to an invalid index while also saving us another search through the logs.
```suggestion
    setTimeout(() => {
      const index = logs.findIndex((l) => l.id === selectedLog.id);
      if (index > -1) {      
        virtuosoRef.current?.scrollToIndex({
          behavior: 'smooth',
          index,
        });
        setScrolledForSearch(true);
      }
    }, 250);
```

Reply:
Our log viewer component assumes that we start from either the beginning or the end, so when selecting a log, it basically scrolls up until the selected log is reached